watcher.go

```diff
@@ -409,101 +356,256 @@ func (w *Watcher) Stop() {
+// It will require updating though, and complicates things. So wait.
+
 type tracker struct {
-	sync.RWMutex
-	deps depmap
-	tpls depmap
+	sync.Mutex
+	// think in terms of a many-2-many DB relationship
+	tracked []trackedPair
+	// viewID -> view
+	views map[string]*view
+	// stringID -> Notifier (stringID is usually template-id)
+	notifiers map[string]Notifier
 }
 
-func newTracker() *tracker {
-	return &tracker{
-		deps: make(depmap),
-		tpls: make(depmap),
+// viewCount returns the number of views watched
+func (t *tracker) viewCount() int {
+	t.Lock()
+	defer t.Unlock()
+	return len(t.views)
+}
+
+// notUsed clears the inUse flag, for testing
+func (t *tracker) notUsed(notifierID, viewID string) bool {
+	for i, tp := range t.tracked {
+		if tp.view == viewID && tp.notify == notifierID {
+			t.tracked[i] = tp.refresh()
+			return true
+		}
 	}
+	return false
 }
 
-func (t *tracker) update(tmplID string, deps ...dep.Dependency) {
-	t.clear(tmplID)
+// lookup returns the view and true, or view's zero value and false
+func (t *tracker) lookup(viewID string) (*view, bool) {
 	t.Lock()
 	defer t.Unlock()
-	for _, d := range deps {
-		t.deps.add(d.String(), tmplID)
-		t.tpls.add(tmplID, d.String())
+	v, ok := t.views[viewID]
+	return v, ok
+}
+
+// Adds new entry
+func (t *tracker) add(v *view, n Notifier) {
+	t.Lock()
+	defer t.Unlock()
+	t.views[v.ID()] = v
+	t.notifiers[n.ID()] = n
+	t.tracked = append(t.tracked,
+		trackedPair{view: v.ID(), notify: n.ID(), inUse: true})
+}
+
+// Marks all trackedPairs w/ a view as having been used
+func (t *tracker) usedID(viewID string) {
+	t.Lock()
+	defer t.Unlock()
+	for idx, tp := range t.tracked {
+		if tp.view == viewID {
+			t.tracked[idx] = tp.used()
+		}
+	}
+}
+
+// Remove view and all trackedPairs that contained it
+func (t *tracker) remove(viewID string) bool {
+	t.Lock()
+	defer t.Unlock()
+	delete(t.views, viewID)
+	tmp := t.tracked[:0]
+	var removed bool
+	for _, tp := range t.tracked {
+		if tp.view != viewID {
+			tmp = append(tmp, tp)
+		} else {
+			removed = true
+		}
 	}
+	t.tracked = tmp
+	return removed
 }
 
-func (t *tracker) clear(tmplID string) {
+// stop all view from polling/watching
+func (t *tracker) stopViews() {
 	t.Lock()
 	defer t.Unlock()
-	for d := range t.tpls[tmplID] {
-		delete(t.deps[d], tmplID)
-		delete(t.tpls[tmplID], d)
+	for id, view := range t.views {
+		delete(t.views, id)
+		if view == nil {
+			continue
+		}
+		view.stop()
+	}
+}
+
+// Return all views for a notifier
+func (t *tracker) notifiersFor(v *view) []Notifier {
+	viewID := v.ID()
+	results := make([]Notifier, 0, 8)
+	for _, tp := range t.tracked {
+		if tp.view == viewID {
+			results = append(results, t.notifiers[tp.notify])
+		}
 	}
+	return results
+}
+
+// initialized returns true if the view has had its data fetched at least once
+func (t *tracker) initialized(viewID string) bool {
+	return t.views[viewID].receivedData
```

Comment:
Does this work as intended? it reads to me that if `viewID` does not exist in the map, it'll be calling `.receivedData` on `nil`

https://play.golang.com/p/5TCcf3QUlKm

Reply:
Yep. In the original version `views` was a `map[string]view` (non-pointer) which worked fine with the zero value. But then I changed it to `map[string]*view` which broke that. I really don't like pointer's zero value semantics at times.